#include "widget.h"
#include "ui_widget.h"
#include <QHostAddress>
#include <QMessageBox>
#include <QDateTime>

Widget::Widget(QWidget *parent) :
    QWidget(parent),
    ui(new Ui::Widget)
{
    ui->setupUi(this);

    // –°–æ–∑–¥–∞–µ–º –∏ –Ω–∞—Å—Ç—Ä–∞–∏–≤–∞–µ–º –∫–ª–∏–µ–Ω—Å—Ç–∫–∏–π —Å–æ–∫–µ—Ç.
    socket = new QTcpSocket(this);
    connect(socket, &QTcpSocket::readyRead, this, &Widget::socketReadyRead);
    connect(socket, &QTcpSocket::disconnected, this, &Widget::socketDisconnected);
    connect(socket, &QTcpSocket::connected, this, &Widget::socketConnected);
    connect (socket,SIGNAL(error(QAbstractSocket::SocketError)),this,SLOT(socketErr(QAbstractSocket::SocketError)));
}

Widget::~Widget()
{
    delete ui;
}

void Widget::on_connectToServerButton_clicked()
{
    // "–ö–æ–æ—Ä–¥–∏–Ω–∞—Ç—ã" —Å–µ—Ä–≤–µ—Ä–∞ (IP-–∞–¥—Ä–µ—Å –∏ –ø–æ—Ä—Ç).
    QString serverIp = ui->serverIpEdit->text();
    QHostAddress serverAddress(serverIp);
    ushort serverPort = ui->serverPortEdit->text().toUShort();

    // –ü–æ–¥–∫–ª—é—á–∞–µ–º—Å—è –∫ —Å–µ—Ä–≤–µ—Ä—É.
    // –ú–æ–∂–µ—Ç –±—ã—Ç—å –¥–æ–ª–≥–æ...
    socket->connectToHost(serverAddress, serverPort);

    // –î–æ–∂–¥–∞—Ç—å—Å—è –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è –∫ —Å–µ—Ä–≤–µ—Ä—É –º–æ–∂–Ω–æ –∏ —Ç—É—Ç.
    // –û–∂–∏–¥–∞–Ω–∏–µ (10 —Å–µ–∫) –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è ...
    bool connected = socket->waitForConnected(10000);
    ui->sendMessageButton->setEnabled(connected);
    ui->pushButton->setEnabled(connected);
    ui->pushButton_2->setEnabled(connected);
    if(!connected)
    {
        // –ü–æ–¥–∫–ª—é—á–∏—Ç—å—Å—è –∫ —Å–µ—Ä–≤–µ—Ä—É –Ω–µ —É–¥–∞–ª–æ—Å—å (–∑–∞–∫–æ–Ω—á–∏–ª—Å—è —Ç–∞–π–º–∞—É—Ç –æ–∂–∏–¥–∞–Ω–∏—è).
        socket->disconnectFromHost();
        // QMessageBox::warning(this,"–í–Ω–∏–º–∞–Ω–∏–µ!","–ù–µ —É–¥–∞–ª–æ—Å—å –ø–æ–¥–∫–ª—é—á–∏—Ç—å—Å—è –∫ —Å–µ—Ä–≤–µ—Ä—É!");
    }
    //    else
    //    {
    //        socket->write((3 + (ui->nickEdit->text())).toUtf8());
    //    }
}
//–æ—Ç–ø—Ä–∞–≤–∫–∞ —Å–æ–æ–±—â–µ–Ω–∏—è –Ω–∞ —Å–µ—Ä–≤–µ—Ä
void Widget::on_sendMessageButton_clicked()
{
    if (!ui->sendMessageEdit->toPlainText().isEmpty()){
        QMessageBox::warning(this,"–û—à–∏–±–∫–∞","–ü—É—Å—Ç–æ–µ —Å–æ–æ–±—â–µ–Ω–∏–µ.");
    }
    QDataStream out (socket);
    QString nick = ui->nickEdit->text();
    out << qint32(1) << QTime::currentTime() << nick << ui->sendMessageEdit->toPlainText();
    ui->chatEdit->append("ME");
    ui->chatEdit->append(ui->sendMessageEdit->toPlainText());
    ui->sendMessageEdit->clear();

}

// –û–±—Ä–∞–±–æ—Ç–∫–∞ –ø—Ä–∏–µ–º–∞ –¥–∞–Ω–Ω—ã—Ö –æ—Ç —Å–µ—Ä–≤–µ—Ä–∞
void Widget::socketReadyRead()
{
    QDataStream stream (socket);

    while(socket->bytesAvailable() > 0)
    {
        stream.startTransaction();

        int id;
        stream >> id;

        if (id == 3 ) { // –°–µ—Ä–≤–µ—Ä –ø–µ—Ä–µ—Å–ª–∞–ª —Å–æ–æ–±—â–µ–Ω–∏–µ –¥—Ä—É–≥–æ–≥–æ –∫–ª–∏–µ–Ω—Ç–∞
            QTime time;
            QString mes;
            QString nick;
            stream >> time >> nick >> mes;

            if(!stream.commitTransaction()) {
                stream.rollbackTransaction();
                break;
            }

            ui->chatEdit->append(time.toString() + "\n" + nick + "\n" + mes);
        }
        else if(id == 4) { // –°–µ—Ä–≤–µ—Ä –ø—Ä–∏—Å–ª–∞–ª —Å–ø–∏—Å–æ–∫ –≤—Å–µ—Ö –∫–ª–∏–µ–Ω—Ç–æ–≤
            QStringList users;
            stream >> users;

            if(!stream.commitTransaction()) {
                stream.rollbackTransaction();
                break;
            }

            ui->nickList->clear();
            ui->nickList->addItems(users);
        }
    }
}
// –û–±—Ä–∞–±–æ—Ç–∫–∞ –æ—Ç–∫–ª—é—á–µ–Ω–∏—è –æ—Ç —Å–µ—Ä–≤–µ—Ä–∞
void Widget::socketDisconnected()
{
    ui->sendMessageButton->setEnabled(false);
    ui->connectToServerButton->setEnabled(true);
    ui->nickEdit->setEnabled(true);
}

// –û–±—Ä–∞–±–æ—Ç–∫–∞ –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è –∫ —Å–µ—Ä–≤–µ—Ä—É
void Widget::socketConnected()
{
    // –î–æ–∂–¥–∞—Ç—å—Å—è –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è –∫ —Å–µ—Ä–≤–µ—Ä—É –º–æ–∂–Ω–æ –∏ —Ç—É—Ç.
    ui->sendMessageButton->setEnabled(true);
    QMessageBox::information(this,"–í–Ω–∏–º–∞–Ω–∏–µ","–ü–æ–¥–∫–ª—é—á–µ–Ω–æ –∫ —Å–µ—Ä–≤–µ—Ä—É.");
    QDataStream stream(socket);
    stream << qint32(2) << ui->nickEdit->text();
}
//–æ–±—Ä–∞–±–æ—Ç–∫–∞ –æ—à–∏–±–æ–∫ –ø—Ä–∏ –ø–æ–¥–∫–ª—é—á–µ–Ω–∏–∏
void Widget::socketErr(QAbstractSocket::SocketError err)
{
    QString error = "–û—à–∏–±–∫–∞: "+
            ( err == QAbstractSocket::HostNotFoundError ? "–•–æ—Å—Ç –Ω–µ –Ω–∞–π–¥–µ–Ω." :
              err == QAbstractSocket::RemoteHostClosedError ? "–•–æ—Å—Ç –∑–∞–∫—Ä—ã—Ç." :
              err == QAbstractSocket::ConnectionRefusedError ? "–û—Ç–∫–∞–∑–∞–Ω–æ –≤ —Å–æ–µ–¥–∏–Ω–µ–Ω–∏–∏." :
                QString(socket->errorString())) ;

    QMessageBox::warning(this,"",error);
}

void Widget::on_pushButton_clicked()
{
    ui->sendMessageEdit->insertPlainText("\U0001F602");
}

//–ø–∞—Ä–∞ —ç–º–æ–¥–∑–∏
void Widget::on_pushButton_2_clicked()
{
    ui->sendMessageEdit->insertPlainText("üòé");
}


